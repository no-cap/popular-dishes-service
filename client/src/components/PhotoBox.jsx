import React from 'react';
import $ from 'jquery';
import { PhotoBoxFormat, PhotoImage, LeftArrow, RightArrow, Caption, Caption2, CaptionCropper, Photos } from '../components-style/PhotoBox-style.jsx';

class PhotoBox extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      currentPhoto: 0
    };
    this.onClickHandler = this.onClickHandler.bind(this);
    this.resetCurrent = this.resetCurrent.bind(this);
    this.handleBlackSpace = this.handleBlackSpace.bind(this);
    this.switchPhotosOnKeyPress = this.switchPhotosOnKeyPress.bind(this);
  }

  resetCurrent() {
    this.setState({ currentPhoto: 0 });
  }

  onClickHandler(e) {
    if (e.preventDefault) {
      e.preventDefault();
    }
    if (e.target.className.includes('button right')) {
      if (this.state.currentPhoto === this.props.photos.length - 1) {
        this.setState({ currentPhoto: 0 });
      } else {
        this.setState({ currentPhoto: this.state.currentPhoto + 1 });
      }
    } else {
      if (this.state.currentPhoto === 0) {
        this.setState({ currentPhoto: this.props.photos.length - 1 });
      } else {
        this.setState({ currentPhoto: this.state.currentPhoto - 1 });
      }
    }
  }

  handleBlackSpace(url) {
    var temporaryImage = new Image()
    temporaryImage.src = url;
    $(temporaryImage).one('load', function () {
      var orgWidth = temporaryImage.width;
      var orgHeight = temporaryImage.height;
      if ((orgWidth/orgHeight) > (800/600)) {
        return 'tooTall';
      } 
    });
  }

  componentDidMount() {
    document.addEventListener("keydown", this.switchPhotosOnKeyPress);
  }

  componentWillUnmount() {
    document.removeEventListener("keydown", this.switchPhotosOnKeyPress, false);
  }

  switchPhotosOnKeyPress(e) {
    var rightObj = { target: { className: "button right" } }
    var leftObj = { target: { className: "left" } }
    if (e.keyCode === 39) this.onClickHandler(rightObj);
    if (e.keyCode === 37) this.onClickHandler(leftObj);
  }

  render() {
    return (
      <div>
        {(this.props.photos.length > 0) ?
          <div>
            <PhotoBoxFormat>
                <PhotoImage src={this.props.photos[this.state.currentPhoto].url} size={this.handleBlackSpace(this.props.photos[this.state.currentPhoto].url)}></PhotoImage>
            </PhotoBoxFormat>
            <LeftArrow onClick={this.onClickHandler}/>
            <RightArrow onClick={this.onClickHandler} className="button right"/>
            <CaptionCropper>
              <Caption2>{this.props.photos[this.state.currentPhoto].caption}</Caption2>
            </CaptionCropper>
            <Caption>{this.props.photos[this.state.currentPhoto].caption}</Caption>
            <Photos>{this.state.currentPhoto + 1} of {this.props.photos.length}</Photos>
          </div> :
          <div>
            <PhotoBoxFormat>
                <PhotoImage src='https://www.yorkshirecareequipment.com/wp-content/uploads/2018/09/no-image-available.jpg' className="photo"></PhotoImage>
            </PhotoBoxFormat>
            <LeftArrow onClick={this.onClickHandler}/>
            <RightArrow onClick={this.onClickHandler}/>
            <Photos>0 of 0</Photos>
          </div>
        }
      </div>
    )

  }

}


export default PhotoBox;