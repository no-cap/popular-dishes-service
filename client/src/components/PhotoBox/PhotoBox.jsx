import React from 'react';
import $ from 'jquery';
import {
  PhotoBoxFormat, PhotoImage, LeftArrow, RightArrow, Caption, Caption2, CaptionCropper, Photos,
} from './PhotoBoxStyles';

class PhotoBox extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      currentPhoto: 0,
    };
    this.onClickHandler = this.onClickHandler.bind(this);
    this.resetCurrent = this.resetCurrent.bind(this);
    this.handleBlackSpace = this.handleBlackSpace.bind(this);
    this.switchPhotosOnKeyPress = this.switchPhotosOnKeyPress.bind(this);
  }

  componentDidMount() {
    document.addEventListener('keydown', this.switchPhotosOnKeyPress);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.switchPhotosOnKeyPress, false);
  }

  onClickHandler(e) {
    if (e.preventDefault) {
      e.preventDefault();
    }
    if (e.target.className.includes('button right')) {
      if (this.state.currentPhoto === this.props.photos.length - 1) {
        this.setState({ currentPhoto: 0 });
      } else {
        this.setState({ currentPhoto: this.state.currentPhoto + 1 });
      }
    } else if (this.state.currentPhoto === 0) {
      this.setState({ currentPhoto: this.props.photos.length - 1 });
    } else {
      this.setState({ currentPhoto: this.state.currentPhoto - 1 });
    }
  }

  resetCurrent() {
    this.setState((state) => ({ ...state, currentPhoto: 0 }));
  }

  // eslint-disable-next-line class-methods-use-this
  handleBlackSpace(url) {
    const temporaryImage = new Image();
    temporaryImage.src = url;
    $(temporaryImage).one('load', () => {
      const orgWidth = temporaryImage.width;
      const orgHeight = temporaryImage.height;
      if ((orgWidth / orgHeight) > (800 / 600)) {
        return 'tooTall';
      }
    });
  }

  switchPhotosOnKeyPress(e) {
    const rightObj = { target: { className: 'button right' } };
    const leftObj = { target: { className: 'left' } };
    if (e.keyCode === 39) this.onClickHandler(rightObj);
    if (e.keyCode === 37) this.onClickHandler(leftObj);
  }

  render() {
    return (
      <div>
        {(this.props.photos.length > 0)
          ? (
            <div>
              <PhotoBoxFormat>
                <PhotoImage src={this.props.photos[this.state.currentPhoto]} size={this.handleBlackSpace(this.props.photos[this.state.currentPhoto])} />
              </PhotoBoxFormat>
              <LeftArrow onClick={this.onClickHandler} />
              <RightArrow onClick={this.onClickHandler} className="button right" />
              <CaptionCropper>
                <Caption2>{this.props.photos[this.state.currentPhoto].caption}</Caption2>
              </CaptionCropper>
              <Caption>{this.props.photos[this.state.currentPhoto].caption}</Caption>
              <Photos>
                {this.state.currentPhoto + 1}
                {' '}
                of
                {' '}
                {this.props.photos.length}
              </Photos>
            </div>
          )
          : (
            <div>
              <PhotoBoxFormat>
                <PhotoImage src="https://www.yorkshirecareequipment.com/wp-content/uploads/2018/09/no-image-available.jpg" className="photo" />
              </PhotoBoxFormat>
              <LeftArrow onClick={this.onClickHandler} />
              <RightArrow onClick={this.onClickHandler} />
              <Photos>0 of 0</Photos>
            </div>
          )}
      </div>
    );
  }
}

export default PhotoBox;
