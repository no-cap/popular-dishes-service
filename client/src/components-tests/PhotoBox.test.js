import React from 'react';
import { shallow, mount } from 'enzyme';
import PhotoBox from '../components/PhotoBox.jsx';
import PhotoEntry from '../components/PhotoEntry.jsx';

// data for one multiple photos
var photos = [
    {
        url: 'https://s3-media0.fl.yelpcdn.com/bphoto/EhwQv7W3nmJqdXhvCW7pJg/o.jpg',
        caption: 'Red Dragon roll was so good!! We devoured it in record timing!'
    },
    {
        url: 'https://s3-media0.fl.yelpcdn.com/bphoto/EhwQv7W3nmJqdXhvCW7pJg/o.jpg',
        caption: 'Red Dragon roll was so good!! We devoured it in record timing!'
    },
    {
        url: 'https://s3-media0.fl.yelpcdn.com/bphoto/EhwQv7W3nmJqdXhvCW7pJg/o.jpg',
        caption: 'Red Dragon roll was so good!! We devoured it in record timing!'
    },
    {
        url: 'https://s3-media0.fl.yelpcdn.com/bphoto/EhwQv7W3nmJqdXhvCW7pJg/o.jpg',
        caption: 'Red Dragon roll was so good!! We devoured it in record timing!'
    }
];

// for making a mock event
const mockEventGenerator = (str) => {
    return {
        target: {
            name: 'todo',
            value: str,
        },
        preventDefault: () => { },
    };
};


// all the tests
describe('PhotoBox Unit test', () => {
    // just seeing if the component even renders correctly
    describe('PhotoBox should render correctly', () => {
        test('PhotoBox should render correctly in "debug" mode', () => {

            const wrapper = shallow(<PhotoBox photos={photos} />);

            expect(wrapper).toMatchSnapshot();
        });

        test('PhotoBox should have only 1 PhotoEntry', () => {

            const wrapper = mount(<PhotoBox photos={photos} />);

            expect(wrapper.find('PhotoEntry')).toHaveLength(1);
        });
    })
    // making sure the component's click handler reacts correctly
    describe('PhotoBox Unit Interaction test', () => {
        describe('onClickHandler', () => {
            test('it should invoke the correct method when a change event is emitted', () => {
                const mockClickHandler = jest.fn();
                const wrapper = shallow(<PhotoBox photos={photos} />);
                wrapper.instance().onClickHandler = mockClickHandler;
                wrapper.instance().forceUpdate();
                wrapper.find('button').at(0).simulate('click');
                wrapper.find('button').at(1).simulate('click');
                expect(mockClickHandler).toHaveBeenCalled();
            });
            // testing that click button increase photo count
            test('it should correctly update the state when a change event is emitted', () => {
                const wrapper = shallow(<PhotoBox photos={photos} />);
                // wrapper.find('button').at(0).simulate('click');
                // expect(wrapper.instance().state.currentPhoto).toBe(7);
                wrapper.find('button').at(1).simulate('click', mockEventGenerator('right'));
                expect(wrapper.instance().state.currentPhoto).toBe(1);
                expect(wrapper.find('.photo-number').text()).toEqual('2 of 4');
                wrapper.find('button').at(1).simulate('click', mockEventGenerator('left'));
                expect(wrapper.instance().state.currentPhoto).toBe(0);
                wrapper.find('button').at(1).simulate('click', mockEventGenerator('left'));
                expect(wrapper.instance().state.currentPhoto).toBe(photos.length - 1);
                expect(wrapper.find('.photo-number').text()).toEqual('4 of 4');
                wrapper.find('button').at(1).simulate('click', mockEventGenerator('right'));
                expect(wrapper.instance().state.currentPhoto).toBe(0);
                expect(wrapper.find('.photo-number').text()).toEqual('1 of 4');
            });
        });
    });

})
