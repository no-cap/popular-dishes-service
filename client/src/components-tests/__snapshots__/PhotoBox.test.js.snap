// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoBox Unit test PhotoBox should render correctly PhotoBox should render correctly in "debug" mode 1`] = `ShallowWrapper {}`;
